### ISSUE TITLE
Querying `last` table changes the behavior of the `logged_in_users` table

### ISSUE THREAD

**Issue by mike-myers-tob** [osquery/osquery#7577]
# Bug report

### What operating system and version are you using?

```
+----------------------+-------+----------+
| version              | build | platform |
+----------------------+-------+----------+
| 21.10 (Impish Indri) |       | ubuntu   |
+----------------------+-------+----------+
```

Also observed on RHEL 7. But, just checked, it does _not_ affect macOS 12 so it's probably Linux-specific API behavior.

### What version of osquery are you using?

```
+---------------------+
| version             |
+---------------------+
| 5.2.2-61-gafe66087b |
+---------------------+
```

But this behavior was also observed in a previous release of osquery.

### What steps did you take to reproduce the issue?

- Start osqueryi, and `select COUNT(*) from logged_in_users;`. This should return only a couple of rows because I am running as the only user.
- Now `select COUNT(*) from last;` which, coincidentally, touches some of the same APIs used by the `logged_in_users` table.
- Now `select COUNT(*) from logged_in_users;` again. It is now, and subsequently will be, incorrect. It returns too much.

I suspect this is because the `last` table calls `utmpxname()` to control the behavior of `getutxent()` which both tables use, but `logged_in_users` expects it to be in its default state before using it. A fix would seem to involve reverting this state change by `logged_in_users` after it is done generating the table results.

### What did you expect to see?

```
osquery> select COUNT(*) from logged_in_users;
+----------+
| COUNT(*) |
+----------+
| 3        |
+----------+
osquery> select COUNT(*) from last;
+----------+
| COUNT(*) |
+----------+
| 20       |
+----------+
osquery> select COUNT(*) from logged_in_users;
+----------+
| COUNT(*) |
+----------+
| 3        |
+----------+
```

### What did you see instead?

Querying `logged_in_users` several times to prove the point: it is only the `last` query that messes it up.

```
osquery> select COUNT(*) from logged_in_users;
+----------+
| COUNT(*) |
+----------+
| 3        |
+----------+
osquery> select COUNT(*) from logged_in_users;
+----------+
| COUNT(*) |
+----------+
| 3        |
+----------+
osquery> select COUNT(*) from logged_in_users;
+----------+
| COUNT(*) |
+----------+
| 3        |
+----------+
osquery> select COUNT(*) from last;
+----------+
| COUNT(*) |
+----------+
| 20       |
+----------+
osquery> select COUNT(*) from logged_in_users;
+----------+
| COUNT(*) |
+----------+
| 73       |
+----------+
```

**Comment by Smjert** [comment]
Good catch!

There's also a second problem with these global APIs, which is that also concurrent queries can break (scheduled query + distributed query. But also queries from an extension).

There are some re-entrant versions, glibc only, but I'm not sure how they work since they don't have an equivalent for `utmp(x)name`. macOS would be left out too.
So I think we should also create an interface to those APIs that locks everytime they are accessed, so that's thread safe.

**Comment by mike-myers-tob** [comment]
> There's also a second problem with these global APIs, which is that also concurrent queries can break (scheduled query + distributed query. But also queries from an extension).
> 
> There are some re-entrant versions, glibc only, but I'm not sure how they can work since they don't have an equivalent for `utmp(x)name`. macOS would be left out too. So I think we should also create an interface to those APIs that locks everytime they are accessed, so that's thread safe.

Ok, should we open a separate issue for the concern of thread-safety? I suppose it might happen if osquery runs scheduled queries and distributed queries in concurrent threads?

To clarify the current issue, it's happening in the same thread (in fact, if each query were run in its own thread, this apparently wouldn't even happen. The `utmpxname()` API sets the path for its own thread only.

**Comment by mike-myers-tob** [comment]
@Smjert I found this, implying we do not need to be concerned about thread-safety of these APIs:

> These functions need not be reentrant. A function that is not required to be reentrant is not required to be thread-safe.

https://www.unix.com/man-page/posix/3/getutxline/

**Comment by Smjert** [comment]
I think the issue there is understanding what standard actually each system respect. It's always a bit of a mess.

That been said, at least in the glibc implementation there are locks indeed: 
https://code.woboq.org/userspace/glibc/login/utmpxname.c.html (which calls __utmpname below)
https://code.woboq.org/userspace/glibc/login/utmpname.c.html#__utmpname

But if you see, the file that's opened, is global:
https://code.woboq.org/userspace/glibc/login/utmpname.c.html#31

And even if it's protected by locks when changed or accessed, it can still change between calls.
An example I mashed up together:
```
#include <mutex>
#include <iostream>
#include <condition_variable>

std::mutex utmpx_mutex;

bool init{false};
bool init_again{false};
std::condition_variable cv;

void thread() {

    {
        std::unique_lock<std::mutex> lock(utmpx_mutex);
        if (!init) {
            //auto res = utmpxname(_PATH_WTMP);
            auto res = utmpxname(_PATH_UTMPX);
            std::cout << "utmpx res: " << res << std::endl;
            std::cout << "thread id: " << std::this_thread::get_id() << std::endl;
            init = true;
            cv.wait(lock, []{ return init_again; });
        }
    }

    {
        std::unique_lock<std::mutex> lock(utmpx_mutex);
        if(!init_again) {
            //auto res = utmpxname(_PATH_UTMPX);
            auto res = utmpxname(_PATH_WTMP);
            std::cout << "utmpx res: " << res << std::endl;
            std::cout << "thread id: " << std::this_thread::get_id() << std::endl;
            init_again = true;
            cv.notify_all();
        }
    }

    setutxent();

    struct utmpx* ut;
    while ((ut = getutxent()) != nullptr) { 
       std::cout << "Username: " << ut->ut_user << std::endl;
    }

}

int main() {
    auto t1 = std::thread(thread);
    auto t2 = std::thread(thread);

    t1.join();
    t2.join();

    return 0;
}
```

Basically the first thread that arrives will set one file and wait for the second thread to set the other.
If there would be no issue, each thread will end up reading its own file anyway, but running this you'll see that the second thread will set the file to read for both.

**Comment by mike-myers-tob** [comment]
Does `logged_in_users` table acquiring this mutex help at all? 

https://github.com/osquery/osquery/blob/master/osquery/tables/system/posix/logged_in_users.cpp#L37
